### PR TITLE
11 get api articles topic query

### PR DESCRIPTION
**Pull request by WolfieKnee**
This will add completed ticket 11 to add query by topic on /api/articles endpoint. 
Adds a controllers utility function to check if the requested topic exists.
Return empty array for a valid topic with no articles.
Ignores an invalid query and servers all articles.

### CODE REVIEW

File: controllers/articles.controllers.js [hunk 1]
```diff
@@ -15,13 +16,25 @@ exports.getArticleById = (req, res, next) => {
...
+	if (topic) {
+		const topicExistsQuery = checkTopicExists(topic);
+		queries.push(topicExistsQuery);
+	}
+	Promise.all(queries)
+		.then((resolvedPromises) => {
+			const articles = resolvedPromises[0];
+			articles.forEach((article) => {
+				return delete article.body;
+			});
```
Comment (nc-poonam-rajput) [inline controllers/articles.controllers.js:32]: this should be done with the query. not manually by you

File: models/articles.models.js [hunk 2]
```diff
@@ -17,18 +17,19 @@ exports.fetchArticleById = (article_id) => {
...
-			ORDER BY created_at DESC`
-		)
-		.then((results) => {
-			return results.rows;
-		});
+exports.fetchArticles = (topic) => {
+	const queryValues = [];
+	let queryStr = `SELECT articles.*, COUNT(comments.comment_id) 
+	AS comment_count FROM articles LEFT OUTER JOIN comments 
+	ON comments.article_id = articles.article_id`;
```
Comment (nc-poonam-rajput) [inline models/articles.models.js:24]: body should be ommitted here

File: controllers/articles.controllers.js [hunk 3]
```diff
@@ -15,13 +16,25 @@ exports.getArticleById = (req, res, next) => {
...
-		articles.forEach((article) => {
-			return delete article.body;
+exports.getArticles = (req, res, next) => {
+	const { topic } = req.query;
+	const articlesQuery = fetchArticles(topic);
+	const queries = [articlesQuery];
+	if (topic) {
+		const topicExistsQuery = checkTopicExists(topic);
+		queries.push(topicExistsQuery);
+	}
```
Comment (nc-poonam-rajput) [inline controllers/articles.controllers.js:26]: honestly I think this touches on data validation which means it should be happening in the model

**Review by nc-poonam-rajput — COMMENTED** [review]
Fab work 🎉 great testing

File: controllers/articles.controllers.js (hunk 1 above)
Comment (WolfieKnee) [reply controllers/articles.controllers.js:32]: good point!

File: models/articles.models.js (hunk 2 above)
Comment (WolfieKnee) [reply models/articles.models.js:24]: refactored for specified fields: 
`SELECT articles.article_id, articles.title, articles.topic, articles.author, articles.created_at, articles.votes, articles.article_img_url, COUNT...`

File: controllers/articles.controllers.js (hunk 3 above)
Comment (WolfieKnee) [reply controllers/articles.controllers.js:26]: refactored to move all data validation and associated utils from controllers to models: 
https://github.com/WolfieKnee/be-nc-news/pull/15

**Pull request merged**